runtime-modules/storage/src/lib.rs

```diff
@@ -2737,43 +2838,91 @@ impl<T: Trait> Module<T> {
+        let bag_change = upload_params
+            .as_ref()
+            .map(|params| {
+                // ensure coherent account ids for prize
+                if let Some(deletion_prize) = deletion_prize {
+                    ensure!(
+                        params.deletion_prize_source_account_id == deletion_prize.account_id,
+                        Error::<T>::AccountsNotCoherent,
+                    );
+                }
+                Self::validate_bag_change(params)
+            })
+            .transpose()?;
+
+        // check that fees are sufficient
+        let total_upload_fee = deletion_prize
+            .as_ref()
+            .map_or(Zero::zero(), |del_prize| del_prize.prize)
+            .saturating_add(bag_change.as_ref().map_or(Zero::zero(), |bag_change| {
+                Self::compute_upload_fees(bag_change)
+            }));
+
+        Self::ensure_sufficient_balance_for_upload(
+            deletion_prize
+                .as_ref()
+                .map(|deletion_prize| deletion_prize.account_id.clone()),
+            total_upload_fee,
+        )?;
 
+        Ok(bag_change)
+    }
+
+    fn ensure_sufficient_balance_for_upload(
+        deletion_prize_source_account_id: Option<T::AccountId>,
+        required_balance: BalanceOf<T>,
+    ) -> DispatchResult {
+        let usable_balance = deletion_prize_source_account_id.map_or(Zero::zero(), |account_id| {
+            Balances::<T>::usable_balance(account_id)
+        });
+
+        ensure!(
+            usable_balance >= required_balance,
+            Error::<T>::InsufficientBalance
+        );
         Ok(())
     }
 
     // Validates dynamic bag deletion params and conditions. Returns bag's deletion prize.
     fn validate_delete_dynamic_bag_params(
         dynamic_bag_id: &DynamicBagId<T>,
+        with_objects: bool,
     ) -> Result<Option<BalanceOf<T>>, DispatchError> {
         Self::ensure_dynamic_bag_exists(dynamic_bag_id)?;
 
         let dynamic_bag = Self::dynamic_bag(dynamic_bag_id);
 
-        ensure!(
-            dynamic_bag.objects_number == 0,
-            Error::<T>::CannotDeleteNonEmptyDynamicBag
-        );
-
-        if let Some(deletion_prize) = dynamic_bag.deletion_prize {
+        // deletion prize = bag.deletion_prize + total_objects fees if any
+        let deletion_prize = if !with_objects {
             ensure!(
-                <StorageTreasury<T>>::usable_balance() >= deletion_prize,
-                Error::<T>::InsufficientTreasuryBalance
+                dynamic_bag.objects_number == 0,
+                Error::<T>::CannotDeleteNonEmptyDynamicBag
             );
-        }
+            dynamic_bag.deletion_prize.unwrap_or_else(Zero::zero)
+        } else {
+            dynamic_bag
+                .deletion_prize
+                .unwrap_or_else(Zero::zero)
+                .saturating_add(Self::calculate_data_storage_fee(
+                    dynamic_bag.objects_total_size,
+                ))
```

Comment:
This is wrong. When you delete data objects you don't recieve `storage_fee` back, only the deletion prize. So this needs to be the sum of:
- `dynamic_bag.deletion_prize`, plus:
- the sum all `data_object.deletion_prize` for each `data_object` in `dynamic_bag`

Because `storage_fee` is slashed on `upload_data_objects`, this will cause invalid `InsufficientTreasuryBalance` error when `storage_fee` > the sum described above.

You can reproduce it with:
```
docker-compose up -d joystream-node
./query-node/start.sh
./storage-node-v2/scripts/init-dev-bucket.sh
./storage-node-v2/bin/run leader:update-data-fee -f 10 --dev
yarn joystream-cli content:createChannel -i ./cli/examples/content/CreateChannel.json
yarn joystream-cli content:deleteChannel -c 1 --force
```

Reply:
fixed in f2fbb1a993